lib/KDateRange/KDateCalendar.vue

```diff
@@ -0,0 +1,475 @@
+  import KDateDay from './KDateDay';
+
+  export default {
+    name: 'KDateCalendar',
+    components: {
+      KIconButton,
+      KDateDay,
+    },
+    props: {
+      /**
+       * constrains the selection to after this date, disabling dates prior
+       */
+      firstAllowedDate: {
+        type: Date,
+        default: null,
+      },
+      /**
+       * constrains date selection to before this date, disabling dates after
+       */
+      lastAllowedDate: {
+        type: Date,
+        default: new Date(),
+      },
+      /**
+       * default value of selected start date
+       */
+      selectedStartDate: {
+        type: [Date, null],
+        default: null,
+      },
+      /**
+       * default value of selected end date
+       */
+      selectedEndDate: {
+        type: [Date, null],
+        default: null,
+      },
+      /**
+       *  Locale string for date formatting
+       */
+      dateLocale: {
+        type: String,
+        required: true,
+      },
+    },
+    data() {
+      return {
+        // Set preselected dates if selected start and end dates are given
+        dateRange: {
+          start: this.selectedStartDate ? this.selectedStartDate : null,
+          end: this.selectedStartDate && this.selectedEndDate ? this.selectedEndDate : null,
+        },
+        numOfDays: 7,
+        isFirstChoice: this.selectedStartDate ? true : false,
+        activeMonth: new Date().getMonth() - 1,
+        activeYearStart: new Date().getFullYear(),
+        activeYearEnd: new Date().getFullYear(),
+      };
+    },
+    computed: {
+      /**
+       * returns the day of week of the beginning of the left side calendar month
+       */
+      activeMonthDay() {
+        return new Date(this.activeYearStart, this.activeMonth, 1).getDay();
+      },
+      /**
+       * returns the day of week of the beginnning of the right side month
+       */
+      nextActiveMonthDay() {
+        return new Date(this.activeYearEnd, this.nextActiveMonth, 1).getDay();
+      },
+      /**
+       * returns the last day of the month on the left side calendar month
+       */
+      activeMonthDate() {
+        return new Date(this.activeYearEnd, this.nextActiveMonth, 0).getDate();
+      },
+      /**
+       * returns the last day of the month on the right side calendar month
+       */
+      nextActiveMonthDate() {
+        return new Date(this.activeYearEnd, this.activeMonth + 2, 0).getDate();
+      },
+      /**
+       * returns the next month that comes after the current active month
+       */
+      nextActiveMonth() {
+        return this.activeMonth >= 11 ? 0 : this.activeMonth + 1;
+      },
+      buttonStyles() {
+        return {
+          '--selected-button-bg-color': this.$themeBrand.secondary.v_50,
+          '--disabled-button-color': this.$themePalette.grey.v_300,
+        };
+      },
+    },
+    watch: {
+      nextActiveMonth(value) {
+        if (value === 0) this.activeYearEnd = this.activeYearStart + 1;
```

Comment:
In this case it is because a computed value should not be mutated, referenced [here](https://vuejs.org/guide/essentials/computed.html#best-practices:~:text=a%20computed%20return%20value%20should%20be%20treated%20as%20read%2Donly%20and%20never%20be%20mutated%20%2D%20instead%2C%20update%20the%20source%20state%20it%20depends%20on%20to%20trigger%20new%20computations):

> a computed return value should be treated as read-only and never be mutated - instead, update the source state it depends on to trigger new computations

So when the Previous/Next Month buttons are clicked, the `activeYearStart` and `activeYearEnd` values will be updated directly and that will trigger the computed properties that depend on these values to update as well.

Reply:
Yes, you shouldn't modify data in a computed prop, but that wasn't quite what I was referring to. So why did you choose a watcher instead of adding this code to the handlers for when the Previous/Next Month buttons are clicked? I do see code in those handlers that are modifying `activeYearEnd`, so it isn't clear why both are needed?

With regards to this specific code-- what happens if the picker is on the month of December, and I click Next, then Previous, then Next, then Previous, and so on? Does `this.activeYearEnd` just keep on incrementing? I suspect the code in the button handlers is possibly covering up a potential issue?

Reply:
Okay, I misunderstood your question at first. The watcher was chosen as a convenient way of keeping tabs on when the `nextActiveMonth` becomes 0 (January) because that is when `activeYearEnd` should be updated to the next year. 

When `goNextMonth` is pressed, the `activeMonth` is what is actually being updated, not `activeYearEnd`, so you are right, changing `activeYearEnd` inside of the `goNextMonth` method is  redundant.

`activeYearEnd` only needs to update when `nextActiveMonth` becomes 0 or when `goPrevMonth` is pressed and `activeYearStart` isn't the same value as `activeYearEnd`, which indicates that the right half of the calendar is in a different year. So pressing `goPrevMonth` would put both sides of the calendar in the same year when it updates the month, this is what decrements the `activeYearEnd`.

I thought the watcher was appropriate because `nextActiveMonth` is only changed after `activeMonth` is updated (which happens inside `goNextMonth`), so the watcher allows `activeYearEnd` to be updated only when the computed property `nextActiveMonth` is a certain number. 

But, based on what you're suggesting, I should be able to make `activeYearEnd` and `activeYearStart` computed properties where the original values are set using `activeMonth`, so they should both update if `activeMonth` is updated and I can still check the value of `nextActiveMonth` when updating `activeYearEnd`. I can implement it this way and update you if I have any issues!